src/EdFi.AnalyticsMiddleTier.DataStandard2/Base/MSSQL/0031-View-AcademicTimePeriodDim.sql

```diff
@@ -0,0 +1,82 @@
+﻿-- SPDX-License-Identifier: Apache-2.0
+-- Licensed to the Ed-Fi Alliance under one or more agreements.
+-- The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+-- See the LICENSE and NOTICES files in the project root for more information.
+
+IF EXISTS (SELECT 1 FROM INFORMATION_SCHEMA.VIEWS WHERE TABLE_SCHEMA = 'analytics' AND TABLE_NAME = 'AcademicTimePeriodDim')
+BEGIN
+    DROP VIEW analytics.AcademicTimePeriodDim
+END
+GO
+
+CREATE VIEW [analytics].[AcademicTimePeriodDim]
+AS
+     SELECT
+		FORMATMESSAGE(
+			'%s-%s-%s-%s-%s',
+			CAST(Session.SchoolId as VARCHAR),
+			CAST(Session.SchoolYear as VARCHAR),
+			CAST(Session.TermDescriptorId as VARCHAR),
+			CAST(GradingPeriod.GradingPeriodDescriptorId as VARCHAR),
+			CONVERT(VARCHAR, GradingPeriod.BeginDate, 112)
+		) as AcademicTimePeriodKey,
```

Comment:
I recently learned that SQL Server 2016 introduced an updated version of [FORMATMESSAGE](https://docs.microsoft.com/en-us/sql/t-sql/functions/formatmessage-transact-sql?view=sql-server-ver15) allowing us to pass in any string for formatting. I think this is easier to read than the CONCAT statements that were used before, because you can see on line 16 exactly what the format is.

In PostgreSQL it is [FORMAT](https://www.postgresqltutorial.com/postgresql-format/) instead of `FORMATMESSAGE` and otherwise works identically.

Reply:
... and I just learned that there's an important difference between `CONCAT` and `FORMATMESSAGE`: `CONCAT` will automatically replace a `null` value with an empty string, but `FORMATMESSAGE` does not do so and will throw an error. Therefore `COALESCE()` or `ISNULL()` is necessary for nullable columns.